### ISSUE TITLE
Most charms will not be built w/o the rust compiler

### ISSUE THREAD

**Issue by DnPlas** [canonical/bundle-kubeflow#989]
### Bug Description

Recently the dependency on this toolchain was added via packages like `cosl` or libraries like the `k8s-service-info`, though this can happen to any charm that depends on a package that uses the rust compiler to be built from source.

If the rust compiler is not added, `charmcraft pack` will fail with `error: can't find Rust compiler`.

To avoid this issue, the following `build-packages` have to be added to the `charmcraft.yaml` file:

```
build-packages: [cargo, rustc, pkg-config, libffi-dev, libssl-dev]
```

### To Reproduce

Build any CKF charm from `main` or re-try the CI

### Environment

Same as CI

### Relevant Log Output

```shell
::    :: Collecting pydantic-core==2.20.1 (from pydantic->cosl)
::    ::   Downloading pydantic_core-2.20.1.tar.gz (388 kB)
::    ::      ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 388.4/388.4 kB 59.3 MB/s eta 0:00:00
::    ::   Installing build dependencies: started
::    ::   Installing build dependencies: finished with status 'error'
::    ::   error: subprocess-exited-with-error
::    ::
::    ::   × pip subprocess to install build dependencies did not run successfully.
::    ::   │ exit code: 1
::    ::   ╰─> [61 lines of output]
::    ::       Collecting maturin<2,>=1
::    ::         Downloading maturin-1.7.0.tar.gz (188 kB)
::    ::            ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 188.3/188.3 kB 34.9 MB/s eta 0:00:00
::    ::         Installing build dependencies: started
::    ::         Installing build dependencies: finished with status 'done'
::    ::         Getting requirements to build wheel: started
::    ::         Getting requirements to build wheel: finished with status 'done'
::    ::         Preparing metadata (pyproject.toml): started
::    ::         Preparing metadata (pyproject.toml): finished with status 'done'
::    ::       Collecting typing-extensions!=4.7.0,>=4.6.0
::    ::         Using cached typing_extensions-4.12.2-py3-none-any.whl
::    ::       Collecting tomli>=1.1.0 (from maturin<2,>=1)
::    ::         Using cached tomli-2.0.1-py3-none-any.whl
::    ::       Building wheels for collected packages: maturin
::    ::         Building wheel for maturin (pyproject.toml): started
::    ::         Building wheel for maturin (pyproject.toml): finished with status 'error'
::    ::         error: subprocess-exited-with-error
::    ::
::    ::         × Building wheel for maturin (pyproject.toml) did not run successfully.
::    ::         │ exit code: 1
::    ::         ╰─> [34 lines of output]
::    ::             running bdist_wheel
::    ::             running build
::    ::             running build_py
::    ::             creating build
::    ::             creating build/lib.linux-x86_64-cpython-38
::    ::             creating build/lib.linux-x86_64-cpython-38/maturin
::    ::             copying maturin/__main__.py -> build/lib.linux-x86_64-cpython-38/maturin
::    ::             copying maturin/__init__.py -> build/lib.linux-x86_64-cpython-38/maturin
::    ::             running egg_info
::    ::             creating maturin.egg-info
::    ::             writing maturin.egg-info/PKG-INFO
::    ::             writing dependency_links to maturin.egg-info/dependency_links.txt
::    ::             writing requirements to maturin.egg-info/requires.txt
::    ::             writing top-level names to maturin.egg-info/top_level.txt
::    ::             writing manifest file 'maturin.egg-info/SOURCES.txt'
::    ::             reading manifest file 'maturin.egg-info/SOURCES.txt'
::    ::             reading manifest template 'MANIFEST.in'
::    ::             /tmp/pip-build-env-1fj3p7m2/overlay/lib/python3.8/site-packages/setuptools/config/_apply_pyprojecttoml.py:79: SetuptoolsWarning: `install_requires` overwritten in `pyproject.toml` (dependencies)
::    ::               corresp(dist, value, root_dir)
::    ::             warning: no files found matching '*.json' under directory 'src/python_interpreter'
::    ::             writing manifest file 'maturin.egg-info/SOURCES.txt'
::    ::             running build_ext
::    ::             running build_rust
::    ::             error: can't find Rust compiler

# It may come from a different package, though.
```

### Additional Context

[Example](https://github.com/canonical/knative-operators/actions/runs/10039174462/job/27742567350#step:4:1147) where this is happening.

**Comment by DnPlas** [comment]
I have found something interesting, though. The charms that are failing now are building `cosl==0.0.13`, which is inconsistent with the pinned version (e.g. [argo](https://github.com/canonical/argo-operators/blob/main/charms/argo-controller/requirements.txt#L20), [knative](https://github.com/canonical/knative-operators/blob/main/charms/knative-operator/requirements.txt#L20)). This behaviour is odd, and we should look at why the pin is ignored at build time.

This strange behaviour explains why the CI was not failing last week, as in a latest version of `cosl` (0.0.13 and 0.0.14) released just last week and today respectively, `pydantic` was introduced as a [dependency](https://github.com/canonical/cos-lib/blob/main/pyproject.toml#L16).

**Comment by DnPlas** [comment]
For reference also look at https://github.com/canonical/charmcraft/issues/1135#issuecomment-2244878420

**Comment by DnPlas** [comment]
The issue with the inconsistent `cosl` versions seems to be fixed in charmcraft 3.1.0, the version in latest/candidate that we use in all CIs. I have tested it locally and the error does not seem to happen any more, BUT now charms are hitting https://github.com/canonical/bundle-kubeflow/issues/883. As suggested by @misohu the fix for it now is to bump `jinja2` to the latest possible version.

Based on the above then I will proceed with:

1. Bumping the version of `jinja2` in all repositories
2. No action required for bumping the charmcraft version as our CI already uses latest/candidate

This issue should be closed once we confirm that all CIs are running and potentially we should revert the commits that already introduced `rustc` and other dependencies as they should not be required.

**Comment by DnPlas** [comment]
It was confirmed that charmcraft 3.x fixed the issue with PYDEPS overriding the versions of pkgs in requirements.txt
After applying the changes suggested in my previous comment, I can confirm all CIs are not presenting this issue. Closing it as there is no further action.
